services/placement/algo/mirrored.go

```diff
@@ -0,0 +1,299 @@
+			return nil, err
+		}
+
+		if mirrorPlacement, _, err = addInstanceToPlacement(ph.GeneratePlacement(nonEmptyOnly), leavingInstance, false); err != nil {
+			return nil, err
+		}
+	}
+	return placementFromMirror(mirrorPlacement, nonLeavingInstances(p.Instances()), p.ReplicaFactor())
+}
+
+func (a mirroredAlgorithm) AddInstances(
+	p services.Placement,
+	addingInstances []services.PlacementInstance,
+) (services.Placement, error) {
+	if !p.IsMirrored() {
+		return nil, errMirrorAlgoOnNotMirroredPlacement
+	}
+
+	for _, instance := range addingInstances {
+		if _, exist := p.Instance(instance.ID()); exist {
+			// TODO(cw): Should we allow adding a leaving instance?
+			// What if leaving instance paired with new instances?
+			return nil, fmt.Errorf("instance %s already exist in the placement", instance.ID())
+		}
+	}
+
+	mirrorPlacement, err := mirrorFromPlacement(p)
+	if err != nil {
+		return nil, err
+	}
+	mirrorInstances, err := groupInstancesByShardSetID(addingInstances, p.ReplicaFactor())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, instance := range mirrorInstances {
+		if _, ok := mirrorPlacement.Instance(instance.ID()); ok {
+			return nil, fmt.Errorf("shard set id %s already exist in current placement", instance.ShardSetID())
+		}
+		ph := newAddInstanceHelper(mirrorPlacement, instance, a.opts)
+		ph.AddInstance(instance)
+		mirrorPlacement = ph.GeneratePlacement(nonEmptyOnly)
+	}
+
+	return placementFromMirror(mirrorPlacement, nonLeavingInstances(append(p.Instances(), addingInstances...)), p.ReplicaFactor())
+}
+
+func (a mirroredAlgorithm) ReplaceInstance(
+	p services.Placement,
+	instanceID string,
+	addingInstances []services.PlacementInstance,
+) (services.Placement, error) {
+	if !p.IsMirrored() {
+		return nil, errMirrorAlgoOnNotMirroredPlacement
+	}
+
+	if len(addingInstances) != 1 {
+		return nil, fmt.Errorf("invalid number of instances: %d for mirrored replace", len(addingInstances))
+	}
+	addingInstance := placement.CloneInstance(addingInstances[0])
+
+	// Clean up shard states before the replacement operation.
+	mirror, err := mirrorFromPlacement(p)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err = placementFromMirror(mirror, nonLeavingInstances(p.Instances()), p.ReplicaFactor())
+	if err != nil {
+		return nil, err
+	}
+
+	leavingInstance, ok := p.Instance(instanceID)
+	if !ok {
+		return nil, fmt.Errorf("instance %s not found in placement", instanceID)
+	}
+
+	shards := leavingInstance.Shards()
+	for _, s := range shards.All() {
+		// NB(cw): Mark Available as Initializing in case we use shard state
+		// to indicate cutover/cutoff time for the shard.
+		if s.State() == shard.Available {
+			s.SetState(shard.Initializing)
+		}
+	}
+
+	addingInstance = addingInstance.SetShards(shards).SetShardSetID(leavingInstance.ShardSetID())
+
+	// NB(cw): Replace is supposed to be used when the leaving node is considered dead,
```

Comment:
I was trying to keep the SourceID in each initializing shard to be the source shardSetID rather than instanceID, so the shards on each instance in the same shardSet are pretty much the same and only modify capacity per shardset. 
But since we are going to support more customized operations I'm going to make the source ids to be instanceIDs, in which case I can just reuse the sharded algorithm's replace() here.